chapters/part3/fairness.md

```diff
@@ -0,0 +1,731 @@
+  "model": {
+    "type": "allennlp.fairness.bias_mitigator_applicator.BiasMitigatorApplicator", 
+    "base_model": {
+      "_pretrained": {
+        "archive_file": "https://storage.googleapis.com/allennlp-public-models/snli-roberta.2021-03-11.tar.gz",
+        "module_path": "",
+        "freeze": false
+      }
+    },
+    "bias_mitigator": {
+      "type": "linear",
+      "bias_direction": {
+        "type": "two_means",
+        "seed_word_pairs_file": "https://raw.githubusercontent.com/tolga-b/debiaswe/4c3fa843ffff45115c43fe112d4283c91d225c09/data/definitional_pairs.json",
+        "tokenizer": {
+          "type": "pretrained_transformer",
+          "model_name": transformer_model,
+          "max_length": 512
+        }
+      }
+    }
+  },
+  "data_loader": {
+    "batch_sampler": {
+      "type": "bucket",
+      "batch_size" : 32
+    }
+  },
+  "trainer": {
+    "num_epochs": 2,
+    "validation_metric": "+accuracy",
+    "learning_rate_scheduler": {
+      "type": "slanted_triangular",
+      "cut_frac": 0.06
+    },
+    "optimizer": {
+      "type": "huggingface_adamw",
+      "lr": 2e-5,
+      "weight_decay": 0.1,
+    }
+  }
+}
+```
+
+You can run this config file with the following CLI command: `allennlp train config.jsonnet --include-package allennlp_models -s /tmp/snli`.
+
+3. After just two epochs of finetuning, we can observe a considerable reduction in gender bias with respect to occupations, as measured by the `NaturalLanguageInference` bias metric on a [large dataset of neutrally-constructed textual entailment sentence pairs](https://storage.googleapis.com/allennlp-public-models/binary-gender-bias-mitigated-snli-dataset.jsonl). Even better, this comes without a noticeable difference in performance on the test set!
+
+To thoroughly evaluate the reduction in binary gender bias with respect to occupations, we will write an `EvaluateBiasMitigation` subcommand:
+
+```python
+"""
+The `evaluate-bias-mitigation` subcommand can be used to
+compare a bias-mitigated, finetuned model with a baseline, pretrained model
+against an SNLI dataset following the format in [On Measuring
+and Mitigating Biased Inferences of Word Embeddings]
+(https://arxiv.org/pdf/1908.09369.pdf) and reports the
+Net Neutral, Fraction Neutral, and Threshold:tau metrics.
+"""
+
+import argparse
+import json
+import logging
+from typing import Any, Dict, Tuple
+from overrides import overrides
+import tempfile
+import torch
+import os
+
+from allennlp.commands.subcommand import Subcommand
+from allennlp.common import logging as common_logging
+from allennlp.common.util import prepare_environment
+from allennlp.fairness.bias_metrics import NaturalLanguageInference
+from allennlp.data import DataLoader
+from allennlp.models.archival import load_archive
+from allennlp.training.util import evaluate
+
+logger = logging.getLogger(__name__)
+
+
+@Subcommand.register("evaluate-bias-mitigation")
```

Comment:
Is this command general enough that we could just ship it?

Reply:
It's not general enough, which is why @AkshitaB suggested that I include it in here. It could also serve as a good example for people to write their own `Subcommand`s.